types/basetypes/dynamic_type.go

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package basetypes
+
+import (
+	"context"
+	"errors"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+// DynamicTypable extends attr.Type for dynamic types. Implement this interface to create a custom DynamicType type.
+type DynamicTypable interface {
+	attr.TypeWithDynamicValue
+
+	// ValueFromDynamic should convert the DynamicValue to a DynamicValuable type.
+	ValueFromDynamic(context.Context, DynamicValue) (DynamicValuable, diag.Diagnostics)
+}
+
+var _ DynamicTypable = DynamicType{}
+
+// TODO: doc
+type DynamicType struct{}
+
+// ApplyTerraform5AttributePathStep applies the given AttributePathStep to the type.
```

Comment:
Perhaps it's not appropriate here, but is it worth providing additional context on the purpose of the `ApplyTerraform5AttributePathStep` method, such as when it's called, and why?